rpi-4.9.y: vc4 HDMI audio, unlimited CMA memory, bugfix

anholt opened the pull request:
This pull request brings in the HDMI audio support from Boris and myself (you'll need Simon's new .debs for libasound, and to use pulseaudio) that is going to be in 4.12.  It also includes my rework to remove the restriction on CMA maximum size/alignment, which isn't yet reviewed upstream.  Also a small fix for a bug that Stefan Wahren ran into, which is heading to stable.

Commit:
ARM: dts: bcm283x: Add HDMI audio related properties

Add the dmas and dma-names properties to support HDMI audio.

Signed-off-by: Boris Brezillon <boris.brezillon@free-electrons.com>
Signed-off-by: Eric Anholt <eric@anholt.net>
(cherry picked from commit d46d2c6380c10e80e99f6af9067356128bffac6b)

anholt committed:
drm/vc4: Add HDMI audio support

The HDMI encoder IP embeds all needed blocks to output audio, with a
custom DAI called MAI moving audio between the two parts of the HDMI
core.  This driver now exposes a sound card to let users stream audio
to their display.

Using the hdmi-codec driver has been considered here, but MAI meant
having to significantly rework hdmi-codec, and it would have left
little shared code with the I2S mode anyway.

The encoder requires that the audio be SPDIF-formatted frames only,
which alsalib will format-convert for us.

This patch is the combined work of Eric Anholt (initial register setup
with a separate dmaengine driver and using simple-audio-card) and
Boris Brezillon (moving it all into HDMI, massive debug to get it
actually working), and which Eric has the permission to release.

v2: Drop "-audio" from sound card name, since that's already implied
    (suggestion by Boris)

Signed-off-by: Eric Anholt <eric@anholt.net>
Acked-by: Boris Brezillon <boris.brezillon@free-electrons.com>
Link: http://patchwork.freedesktop.org/patch/msgid/20170227202803.12855-2-eric@anholt.net

anholt committed:
drm/vc4: Allow using more than 256MB of CMA memory.

Until now, we've had to limit Raspberry Pi to 256MB of CMA memory to
keep from triggering the hardware addressing bug between of the tile
binner of the tile alloc memory (where the top 4 bits come from the
tile state data array's address).

To work around that and allow more memory to be reserved for graphics,
allocate a single BO to store tile state data arrays and tile
alloc/overflow memory while the GPU is active, and make sure that that
one BO doesn't happen to cross a 256MB boundary.  With that in place,
we can allocate textures and shaders anywhere in system memory (still
contiguous, of course).

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
Revert "drm/vc4: Verify at boot that CMA doesn't cross a 256MB boundary."

This reverts commit 016c804fa69381bd23a1a13cba08a3c4557fcdff.  The
restriction was lifted by the previous commit.

popcornmix commented:
I see the hdmi audio driver only supports IEC958_SUBFRAME_LE
Do we need an alsa config file to support normal PCM formats?

anholt commented:
Yeah, that's the new libasound .debs I mentioned.

popcornmix commented:
I tried installing the alsa deb files @spl237 sent to you:
```
pi@domnfs:/tmp/alsa-lib $ sudo dpkg -i *.deb
(Reading database ... 118209 files and directories currently installed.)
Preparing to unpack libasound2_1.0.28-1+rpi3_armhf.deb ...
Unpacking libasound2:armhf (1.0.28-1+rpi3) over (1.0.28-1+rpi2) ...
Preparing to unpack libasound2-data_1.0.28-1+rpi3_all.deb ...
Unpacking libasound2-data (1.0.28-1+rpi3) over (1.0.28-1+rpi2) ...
Selecting previously unselected package libasound2-dbg:armhf.
Preparing to unpack libasound2-dbg_1.0.28-1+rpi3_armhf.deb ...
Unpacking libasound2-dbg:armhf (1.0.28-1+rpi3) ...
Preparing to unpack libasound2-dev_1.0.28-1+rpi3_armhf.deb ...
Unpacking libasound2-dev:armhf (1.0.28-1+rpi3) over (1.0.28-1+rpi2) ...
Selecting previously unselected package libasound2-doc.
Preparing to unpack libasound2-doc_1.0.28-1+rpi3_all.deb ...
Unpacking libasound2-doc (1.0.28-1+rpi3) ...
Setting up libasound2-data (1.0.28-1+rpi3) ...
Setting up libasound2-doc (1.0.28-1+rpi3) ...
Setting up libasound2:armhf (1.0.28-1+rpi3) ...
Setting up libasound2-dbg:armhf (1.0.28-1+rpi3) ...
Setting up libasound2-dev:armhf (1.0.28-1+rpi3) ...
Processing triggers for libc-bin (2.19-18+deb8u7) ...
```
and it didn't seem to fix the issue:
```
pi@domnfs:~ $ aplay dell/audio/norah.wav 
Playing WAVE 'dell/audio/norah.wav' : Signed 16 bit Little Endian, Rate 44100 Hz, Stereo
aplay: set_params:1233: Sample format non available
Available formats:
- IEC958_SUBFRAME_LE
```
Should this be working?

anholt commented:
Note that at the top of the PR, I said that pulseaudio is what works.  It's going to take some experimenting with the ordering of the plugins, probably, to get basic aplay to work.  Sadly, alsa has atrocious documentation of its config file system.

spl237 commented:
pulseaudio is incompatible with a couple of our major education applications (Scratch and Sonic Pi), which is why I've spent the last few days removing our current dependency on pulseaudio for Bluetooth; we're not planning to ship pulseaudio in the standard image once we move to stretch. We're going to have unhappy education users if we can't get HDMI audio working without pulseaudio. Is this a "only works with pulseaudio at the moment", or is it "only works with pulseaudio, full stop"?

anholt commented:
It's going to take experimenting with the ordering of plugins, but is in principle totally doable.

anholt committed:
panel-raspberrypi-touchscreen: Fix NULL deref if probe order goes wrong.

If the i2c driver hadn't pobed before the panel driver probes, then
the client would be NULL but we were looking for an ERR_PTR in the
error case.

Signed-off-by: Eric Anholt <eric@anholt.net>

ednl commented:
@spl237 Yay! Because in my experience it also interferes with mpd which I always use, and which I fix with "apt-get purge pulseaudio". And I am curious about your question which I think @anholt didn't really answer: does it or does it not break alsa hdmi audio right now? (From the answer I guess it does, but then why would @popcornmix just throw it into the update?)

popcornmix commented:
This PR makes no difference to audio support in my testing.
If you aren't using the vc4 kms driver then there is no change.
If you are using the vc4 kms driver and have "dtparam=audio=on" in config.txt you'll still get firmware driven audio.
If you are using the vc4 kms driver and do not have "dtparam=audio=on" in config.txt you'll get an arm side hdmi audio driver which may not work out of the box, but that wouldn't have worked before. However now there is a possibility that you can test the arm side hdmi audio driver with the correct configuration.

ednl commented:
@popcornmix Thanks for the clarification.
